fuel-relayer/src/service.rs

```diff
@@ -1,67 +1,197 @@
+    pub receiver: mpsc::Receiver<RelayerRequest>,
+    /// Private.
+    pub private_key: Vec<u8>,
+    /// Db connector to apply stake and token deposit.
+    pub db: Box<dyn RelayerDb>,
+    /// Notification of new block event.
+    pub new_block_event: broadcast::Receiver<ImportBlockBroadcast>,
+    /// Relayer Configuration.
+    pub config: Config,
+}
+
+impl Service {
+    pub(crate) fn new(sender: relayer::Sender, context: Context) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            sender,
+            join: Mutex::new(None),
+            context: Arc::new(Mutex::new(Some(context))),
+        })
     }
 
     /// create provider that we use for communication with ethereum.
-    pub async fn provider(uri: &str) -> Result<Provider<Ws>, Error> {
+    pub(crate) async fn provider_ws(uri: &str) -> Result<Provider<Ws>, Error> {
         let ws = Ws::connect(uri).await?;
         let provider =
             Provider::new(ws).interval(std::time::Duration::from_millis(PROVIDER_INTERVAL));
         Ok(provider)
     }
 
-    pub fn provider_http(uri: &str) -> Result<Provider<Http>, Error> {
+    pub(crate) fn provider_http(uri: &str) -> Result<Provider<Http>, Error> {
         let url = Url::parse(uri).unwrap();
         let ws = Http::new(url);
         let provider =
             Provider::new(ws).interval(std::time::Duration::from_millis(PROVIDER_INTERVAL));
         Ok(provider)
     }
 
-    pub fn sender(&self) -> &mpsc::Sender<RelayerEvent> {
+    // internal function to simplify code for different prividers.
+    async fn run<P>(
+        &self,
+        join: &mut Option<JoinHandle<Context>>,
+        context: Context,
+        provider: Result<P, anyhow::Error>,
+    ) -> Result<(), anyhow::Error>
+    where
+        P: Middleware<Error = ProviderError> + 'static,
+    {
+        let provider = match provider {
+            Ok(provider) => provider,
+            Err(e) => {
+                *self.context.lock().await = Some(context);
+                return Err(e);
+            }
+        };
+        let relayer = Relayer::new(context).await;
+        *join = Some(tokio::spawn(async {
+            relayer.run(Arc::new(provider)).await
+        }));
+        Ok(())
+    }
+
+    pub async fn start(&self) -> Result<(), anyhow::Error> {
+        let mut join = self.join.lock().await;
+        if join.is_none() {
+            if let Some(context) = self.context.lock().await.take() {
+                let uri = context.config.eth_client.clone();
+                if uri.starts_with("http") {
+                    self.run(&mut *join, context, Self::provider_http(&uri))
+                        .await
+                } else if uri.starts_with("ws") {
+                    self.run(&mut *join, context, Self::provider_ws(&uri).await)
+                        .await
+                } else {
+                    *self.context.lock().await = Some(context);
+                    Err(anyhow!("Relayer uclient uri is not http or ws"))
```

Comment:
`uclient` -> `client`

```suggestion
                    Err(anyhow!("Relayer client uri is not http or ws"))
```